src/append.jl

```diff
@@ -0,0 +1,168 @@
+    elseif compress isa Symbol
+        throw(ArgumentError("unsupported compress keyword argument value: $compress. Valid values include `:lz4` or `:zstd`"))
+    end
+
+    open(file, "r+") do io
+        append(io, tbl, largelists, compress, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    end
+
+    return file
+end
+
+function append(io, source, largelists, compress, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    bytes = Mmap.mmap(io)
+    stream = Arrow.Stream(bytes)
+    arrow_schema = Tables.schema(Tables.columns(stream))
+    dictencodings = stream.dictencodings
+    firstcols = toarrowtable(stream, dictencodings, largelists, compress, denseunions, dictencode, dictencodenested, maxdepth)
+    sch = Ref{Tables.Schema}(Tables.schema(firstcols))
+
+    # position file at the last footer
+    offset = footer_offset(bytes)
+    seek(io, length(bytes) - offset)
+
+    # TODO: we're probably not threadsafe if user passes own single compressor instance + ntasks > 1
+    # if ntasks > 1 && compres !== nothing && !(compress isa Vector)
+    #     compress = Threads.resize_nthreads!([compress])
+    # end
+    msgs = OrderedChannel{Message}(ntasks)
+    # build messages
+    blocks = (Block[], Block[])
+    # start message writing from channel
+    threaded = ntasks > 1
+    tsk = threaded ? (Threads.@spawn for msg in msgs
+        Base.write(io, msg, blocks, sch, alignment)
+    end) : (@async for msg in msgs
+        Base.write(io, msg, blocks, sch, alignment)
+    end)
+    anyerror = Threads.Atomic{Bool}(false)
+    errorref = Ref{Any}()
+    @sync for (i, tbl) in enumerate(Tables.partitions(source))
+        if anyerror[]
+            @error "error writing arrow data on partition = $(errorref[][3])" exception=(errorref[][1], errorref[][2])
+            error("fatal error writing arrow data")
+        end
+        @debug 1 "processing table partition i = $i"
+        tbl_schema = Tables.schema(tbl)
+
+        if !is_equivalent_schema(arrow_schema, tbl_schema)
+            throw(ArgumentError("Table schema does not match existing arrow file schema"))
+        end
+
+        if threaded
+            Threads.@spawn process_partition(tbl, dictencodings, largelists, compress, denseunions, dictencode, dictencodenested, maxdepth, msgs, alignment, i, sch, errorref, anyerror)
+        else
+            @async process_partition(tbl, dictencodings, largelists, compress, denseunions, dictencode, dictencodenested, maxdepth, msgs, alignment, i, sch, errorref, anyerror)
+        end
+    end
+    if anyerror[]
+        @error "error writing arrow data on partition = $(errorref[][3])" exception=(errorref[][1], errorref[][2])
+        error("fatal error writing arrow data")
+    end
+    # close our message-writing channel, no further put!-ing is allowed
+    close(msgs)
+    # now wait for our message-writing task to finish writing
+    wait(tsk)
+
+    b = FlatBuffers.Builder(1024)
```

Comment:
we shouldn't need any of this code; it's not going to be valid to append to a feather formatted file, which needs this footer stuff.

Reply:
Can't we append to a feather/arrow ipc file format by starting to write from the footer offset? Then we need to write the new footer back at the end, no?

Reply:
Hmmm, that sounds quite a bit more complicated. We'd have to _read_ the existing footer, because we need all that existing info, then we'd need to write out the new footer; and yes, rewrite over the existing one. I think that could be a separate PR, where we first start with only support appending to IPC streams.

Reply:
Actually, appending to stored files is the functionality I need and that is what I tried to address here. But looks like it is valid to store arrow data into files in stream format? Then we don't need to deal with the footer which will make things simpler. Is that what you meant here?

Reply:
done that way in https://github.com/JuliaData/Arrow.jl/pull/160/commits/7aca33da479fad8e98d1ff192d98adf9aba6e8b7 now

Reply:
Yes, that's what I meant. It's much simpler to append to file when data is in IPC format instead of "file" format.